### PR TITLE
Private feed implementation

### LINKED ISSUE #142: Support private feeds
### Describe the solution you'd like
https://github.com/loic-sharma/BaGet/issues/36

### Describe alternatives you've considered
Didnt find any open source nuget with private feeds

### PR DESCRIPTION
**Pull request by tomasfil**
Summary of the changes (in less than 80 chars)

Implements private feeds resolves #142

Used by nuget:
NuGet Sources Add -Name "feed_name" -Source "http://localhost:50557/v3/index.json" -UserName "username" -Password "password"

Addresses https://github.com/bagetter/BaGet/issues/142 https://github.com/bagetter/BaGet/issues/147

### CODE REVIEW

**Comment by tomasfil** [comment]
Added code to resolve #147

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (Regenhardt) [inline src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:58]: Wouldn't this fail the authentication when a package is pushed using the API key?

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (Regenhardt) [inline src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:77]: Please wrap this in a private method like `CreateClaimsFor(user)` or something.

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (Regenhardt) [inline src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:43]: I'd prefer a private method to wrap this. Not sure what to call it, maybe `CreateAnonymousAuthClaim()`? `CreateAnonymousLogin()`? `AcceptAnonymousAuthentication()`? I'd probably consult chatgpt for this one but maybe you have a better idea.

File: src/BaGetter.Web/Controllers/ServiceIndexController.cs
Comment (Regenhardt) [inline src/BaGetter.Web/Controllers/ServiceIndexController.cs:16]: Do these work with how we configure the routes in the `BaGetterEndpointBuilder`?

File: src/BaGetter.Core/Configuration/BaGetterOptions.cs
Comment (Regenhardt) [inline src/BaGetter.Core/Configuration/BaGetterOptions.cs:15]: These just add additional API keys the admin can give to other teams in order to late be able to revoke specific keys, right? Or did I miss something?

**Review by Regenhardt — CHANGES_REQUESTED** [review]
Also I'd like additional opinions on this. Once we have user accounts, we should probably add the user as package owner, even if it's just to see who pushed what package. After that we can talk about authorization concepts like restricting uploads to an owner or something.

File: src/BaGetter.Core/Configuration/AuthenticationOptions.cs
Comment (tomasfil) [reply src/BaGetter.Core/Configuration/AuthenticationOptions.cs:7]: Well. Thats given you would like to implement login screen, user management screen etc. This was intended to serve purpose to of deploying the Bagetter with (docker compose, or appsettings on IIS) and with preset Username/Password (like with api key), to prevent unauthorized download.

That doesnt mean that could not be extended in the future to work for both (this as system/default users and additional user profiles created later on and stored in DB). But that was not t

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (tomasfil) [reply src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:58]: No, I have tested that. I did decorate only controllers that handle downloads and browsing. Uploading is not affected. You can even add -username and -password to nuget push.

File: src/BaGetter.Web/Controllers/ServiceIndexController.cs
Comment (tomasfil) [reply src/BaGetter.Web/Controllers/ServiceIndexController.cs:16]: Yes

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (Regenhardt) [reply src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:49]: Ok I see. Can this be used for the razor pages too?

File: src/BaGetter.Core/Configuration/AuthenticationOptions.cs
Comment (Regenhardt) [reply src/BaGetter.Core/Configuration/AuthenticationOptions.cs:7]: Hm that makes sense, and this is way easier than implementing whole oauth workflows. Any idea if we can do this without the passwords? Can't give the admin your actual password.   
Turns out I was stuck on semantics. Could just call it a personal API key and I'm happy. Unfortunately though nuget calls it password so naturally, writing it down or sharing it feels wrong. 

Make sure to add it to the documentation, then let's see if we can find some more people with opinions on this.

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (tomasfil) [reply src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:49]: Not directly like that. It would need to be implemented differently

File: src/BaGetter.Core/Configuration/AuthenticationOptions.cs
Comment (tomasfil) [reply src/BaGetter.Core/Configuration/AuthenticationOptions.cs:7]: I was thinking about not calling it directly username and password, but the Nuget CLI is the way it is and that would be more confusing than not calling it that way...

For example with telerik private feeds you need to use your actual username and password when you set up the feed. The admin will just "make up" the username and password. 

Then I think next logical step is to open config/way, to let people implement their own auth using maybe HTTP call, or smthing. But that will be much harder than this

Comment (tomasfil) [reply src/BaGetter.Core/Configuration/AuthenticationOptions.cs:7]: I have updated the documentation on private feeds

Comment (tomasfil) [reply src/BaGetter.Core/Configuration/AuthenticationOptions.cs:7]: Also for `Any idea if we can do this without the passwords? Can't give the admin your actual password.`
The username would lean to be "quessable", I think its much better to be provided with memorizeable username, like domain username for PC and then randomly generated password, same with the api key.

But calling it username/apiKey and then setup username/password in nuget CLI seems counter intuitive

File: src/BaGetter.Core/Configuration/BaGetterOptions.cs
Comment (Regenhardt) [inline src/BaGetter.Core/Configuration/BaGetterOptions.cs:68]: I'd prefer this to be an `AuthenticationOptions` object that contains a `Credential[] Credentials` or something, I think that would make it easier to modify later with less breaking changes.

File: src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs
Comment (Regenhardt) [inline src/BaGetter/Authentication/NugetBasicAuthenticationHandler.cs:88]: There's weird formatting going on in this file, please format the code.

**Review by Regenhardt — CHANGES_REQUESTED** [review]
Could you also add an integration test for a private feed? Maybe a new class `AuthIntegrationTests`. Or add an authenticated request to the nuget client integration test, although I'm not sure if it gets ugly to add authentication for just one test method and not the others.

**Comment by tomasfil** [comment]
@Regenhardt LMK if you find anything else with it.

**Comment by tomasfil** [comment]
LMK if you need anything else

File: src/BaGetter/Startup.cs
Comment (Regenhardt) [inline src/BaGetter/Startup.cs:36]: If possible, find out why the test tries to initialize this multiple times and fix it there, I'd like to avoid test code in the application code.

File: tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs
Comment (Regenhardt) [inline tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs:1]: I think both the handler and this test would be better positioned in `BaGetter.Web` and `BaGetter.Web.Tests`, as the `BaGetter` project should mostly be configuration and tying things together. In `BaGetter.Tests` I'd like an integration tests that spins up the `BaGetter` application, if possible. If that's too much for now don't worry about the integration tests though, I can do that too eventually (i.e. when my bachelor's thesis is done).

**Review by Regenhardt — CHANGES_REQUESTED** [review]
Very nice, I think it's just minor things now but all in all it looks good. Would still prefer an additional review though, @bagetter/maintainers. I won't block the PR for long if nobody else finds time for a review either though, don't worry.

File: tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs
Comment (Regenhardt) [reply tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs:1]: Plus I gotta look into that later but I think the handler being in `.Web` will also pave the way to guarding the pages with basic authentication too, if configured. I will learn about that eventually.

**Comment by Regenhardt** [comment]
Oh when the handler is in `.Web`, you could add an extension method so the startup can call it like `services.AddAuthentication().AddNugetBasicHttpAuthentication()` or something. 

Possible calls:  
`services.AddAuthentication().AddNugetBasicHttpAuthentication()` as auth scheme next to calls like ´AddBearerToken()`
`services.AddBasicHttpAuthentication()` as complete alternative to `AddAuthentication()`
`BaGetterApplication.AddBasicHttpAuthentication()` which would make sure it can only used with BaGetter and not with other applications accidentally that just include BaGetter

I'm not currently fit enough to completely make the choice so, anything with an explanation why would be great.

File: src/BaGetter/Startup.cs
Comment (tomasfil) [reply src/BaGetter/Startup.cs:36]:
```
WebApplicationFactory behavior:
When WebApplicationFactory.CreateClient() is called, it creates a new test server, which involves rebuilding the application. This process includes calling ConfigureServices again, which would lead to AddAuthentication being called a second time.
```

Comment (tomasfil) [reply src/BaGetter/Startup.cs:36]: As far I found out, this is propably the cleanest solution around that.

Comment (Regenhardt) [reply src/BaGetter/Startup.cs:36]: Thanks for checking.

File: tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs
Comment (tomasfil) [reply tests/BaGetter.Tests/NugetBasicAuthenticationHandlerTests.cs]: The integration tests are in there already. I did move the handler and handler test as requested

**Comment by tomasfil** [comment]
The apiKey object makes sense. I would split the per/package authentication into separate PR later on. I can implement that. Here I would like to focus on the base with enabling the functionality down the road

**Comment by tomasfil** [comment]
`BaGetterApplication.AddBasicHttpAuthentication()` would require to store extensions in another cs file upper the project scheme, since BaGetter.Core doesnt have auth extensions.

**Comment by tomasfil** [comment]
Yup, makes sense to limit that for BaGetterApplication, so it does not interfere with anything else

**Comment by tomasfil** [comment]
I have a dilema.
Move `ApiKeys` into `NugetAuthenticationOptions`, or leave it in base next to the original `ApiKey`, both have pros and cons.

But maybe moving it into `NugetAuthenticationOptions` would be sensible in case another features like the per package/scope auth feature implementation?

**Comment by Regenhardt** [comment]
I'd say move it into the options object, so the root API key stays as the simple default case where you just want to restrict pushing (I assume many companies just have their feed available in their local network anyway), and the complex options are an optional extension for more sophisticated concepts.

**Comment by tomasfil** [comment]
Will do tomorrow

**Comment by tomasfil** [comment]
As for me its GTG

**Comment by arkangel-dev** [comment]
I propose that we actually implement full access control with actual user entities... Im currently working on implementing that... I've finished linking packages to accounts. It just needs the following
- Implement checks during write operations
- Implement DbMigrations for the storage providers
- Maybe a UI for updating the users. Right now all of it is managed by the REST endpoint

Id love it you went through it @Regenhardt : https://github.com/bagetter/BaGetter/compare/main...arkangel-dev:BaGetter:main

**Comment by tomasfil** [comment]
I understand. But I must say. I come from open-source and corporate. And this implementation is big no go from many point of views, because this means another management and ecosystem locking. 

It propably makes sense to create full access control with user entities, but it is definetly not straighfowrard as this. Because many existing users use Azure, Windows AD auth etc. 

Simple auth scheme such as this can be later on extended, but I dont think its wise to lock users in this way. Because I could see implemening certificate auth, or something completely different like AWS. It should consider authentication flows like OAuth 2.0 etc, to allow automatization of workflows etc... 

I think that the approach should propably be
1. Implement simple and basic configurable auth - Needed by majority
2. Thoroughly design Auth API
3. Allow extension nuget packages to implement and inject them

**Comment by arkangel-dev** [comment]
I just pushed the final commit to my fork, its finished, I understand where you're coming from, what I've done is a very simplified implementation of access control. Just enough to suit my needs and I dont mind if my changes aren't merged into the main fork, but the way I see it, it is bounds better than just multiple api keys having full write access to all packages.

Also i dont understand what you mean by authentication flows like OAuth 2.0... Are you talking about the web page?

**Comment by tomasfil** [comment]
> Also i dont understand what you mean by

> Are you talking about the web page?

What I was trying to say is, that people should have the ability to use the authentication that they have now, if at all possible

**Comment by arkangel-dev** [comment]
I think I got what you mean. The authentication they have now, as in the API key that is currently implemented right?

**Comment by tomasfil** [comment]
No, I think the authentication should be in core simple as possible to allow basic control like this PR does.  - Also package control can be added in similiar fasion

And after that everything else should be abstracted out in some way and possibly provide built in solution like your implementation. But there are many possible solutions from simple to complex and it really depends what are the goals with the Bagetter

**Comment by arkangel-dev** [comment]
I dont mind the authentication. I think its really well done... Its the authorization performed when writing to the package list. I think I'll take a stab at implementing the authorization part after your merge 😄

**Comment by Regenhardt** [comment]
I do like both of your implementations, I think with the extensibility @tomasfil got going on here we could very well extend it using @arkangel-dev 's implementation afterwards, thanks a lot for that! I must admit that, in the beginning, I definitely wanted something like the implementation of @arkangel-dev, but now am excited by the simplicity and extensibility of what @tomasfil built here.

@arkangel-dev How do you actually use your admin controller, do you have your own client? I didn't see a new page for that. Or does the nuget client include an interface for that? There should definitely be a service handling all that from an MVC page, so then there can be both the page and the controller injecting the service.  
Also when checking the token, the checks that do not access the database should probably happen before the check that accesses the database to avoid unnecessary DB connections. Also constants for the auth schemes and roles, unless we secretly plan to make roles configurable too in the future so it doesn't matter anyway.  
I'd be absolutely ok with a draft PR of your changes so we can track it and have a better space to talk about how to proceed with it, even if it takes months. I removed the stale-issue-bot exactly because development is slow here and I don't think a to-do is done just because I ignored it long enough.

@tomasfil I've gotta look at the latest changes but I think we're good to go too, just don't have time right now, I'll try on the weekend.

I do like the idea of another project having the actual identity implementation, we could have BaGetter reference and use that directly and publish it as NuGet package for other people consuming the library. It would however be nice if it's compatible with ASP.NET Core identity somehow, although I just don't have the overview about that one to even guess how compatible it currently it. That would pave the way to easier addition of other identity systems like oauth2 though. 

Similar to how other providers work, the whole identity implementation should only be used (i.e. DB created/migrated and context called) if it is configured at all, if possible the addition should maintain the default configuration of a simple to set up nuget feed with next to no configuration necessary to run. This is of course all speculative about how optional all that can be done. I think BaGetter does have a very nice architecture for these kind of things though.

**Comment by arkangel-dev** [comment]
For my use case, i just make requests manually, but yeah, you should def have a web page. And for the checks, I'll update those. And I'll make a draft after this PR is merged. Thanks!

File: src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs
Comment (Regenhardt) [inline src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs:36]: I'd prefer these calls in the BaGetter project so they are separate from the web app. Technically someone could add BaGetter as nuget feed to their application but not use the web app, in which case they'd have to discover these calls by other means to use them.

File: src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs
Comment (Regenhardt) [inline src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs:59]: If I want to add BaGetter to my application and use multiple policies, would it work with this? Can I `.AddAuthorization()` with my policy and then call this for the nuget feed and they stack? If not, is there a way to make them stack (I can only think of taking an optional `Action<options>` parameter and call it along with `.AddPolicy`.

File: tests/BaGetter.Tests/NugetAuthenticationIntegrationTests.cs
Comment (Regenhardt) [inline tests/BaGetter.Tests/NugetAuthenticationIntegrationTests.cs:67]: This would be nicer as `_client.DefaultRequestHeaders.Authorization = new(...)`, unless it doesn't work in this context?

File: docs/docs/configuration.md
Comment (Regenhardt) [inline docs/docs/configuration.md:153]: I feel like we can assume people already have nuget/dotnet/VS installed and this would be very nice as tabs |NuGet CLI|dotnet CLI|Visual Studio| but I'm not actually sure. I might also accept it like this.

**Review by Regenhardt — CHANGES_REQUESTED** [review]

**Comment by tomasfil** [comment]
I will pop into those requests propably next week. Or anyone interested can really

**Comment by hoerup** [comment]
That is the current status of this ?

**Comment by seriouz** [comment]
@hoerup The last thing I know was that @tomasfil wanted to have a look at the change requests.

**Comment by tomasfil** [comment]
I am sorry, currrently very busy. I will try my best to work on the requested changes this week.

**Comment by seriouz** [comment]
@tomasfil Okay, but don't stress yourself!

**Comment by seriouz** [comment]
@tomasfil  Just bumbing as not to be forgotten :)

File: src/BaGetter.Core/Configuration/ApiKey.cs
Comment (tomasfil) [reply src/BaGetter.Core/Configuration/ApiKey.cs]: Just my own deformation, missed that

File: src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs
Comment (tomasfil) [reply src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs]: Moved

File: src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs
Comment (tomasfil) [reply src/BaGetter.Web/Extensions/IServiceCollectionExtensions.cs]: Moved

**Comment by tomasfil** [comment]
@seriouz @Regenhardt Resolved the comments.

**Comment by Regenhardt** [comment]
Thanks, gonna take a look when I'm back. I've been in the hospital for a week unfortunately.

**Comment by Regenhardt** [comment]
Oh yes, way better, walking almost looks normal again 😊. Still getting settled and cleaning up after a month of not being at home, I'm gonna look at this soon, haven't forgotten it don't worry.

**Comment by chenzuo** [comment]
I hope the PR merge can be successful and add a private package option. for any teams using this open source project hope for authenticated access rather than open access. this is a very important feature. without this feature, we can only solve it through other additional ways, but this solution does not seem to be based on software engineering principles. so I hope  this PR merge successful soon.

**Comment by tomasfil** [comment]
I will just add, that I am using locally deployed image for production sice Jul with 5 developers and everything works as expected, nothing broke yet and we didn't find any security issue of leaking the packages without authorization

**Review by Regenhardt — APPROVED** [review]
LGTM, thanks a lot for this great addition @tomasfil! Will try to make a new release today :)

**Review by Regenhardt — CHANGES_REQUESTED** [review]
Turns out the test project doesn't build, probably needs a using updated.

**Comment by tomasfil** [comment]
The build was broken by https://github.com/tomasfil/BaGetter/commit/0fe5bac368d891883cf18b4b473909603ff495fb

I did fix it.

**Comment by chenzuo** [comment]
I’m waiting for your feature to be released because I plan to share it at a local tech meetup, focusing on project and code security. I’m really excited about it and check multiple times a day to see if it’s out yet.

**Review by Regenhardt — APPROVED** [review]

**Comment by chenzuo** [comment]
thanks a ton for all ! this given us an amazing and super useful feature. we really appreciate it!

**Comment by theazgra** [comment]
Hi,

great to see private feed getting implemented. I have already deployed it. Only question I have is. While feed is private (access to index.json is protected by basic auth) the Web UI search is not limited, and packages can be downloaded from there. Is there solution to this, which I have overlooked?

Thanks

**Comment by tomasfil** [comment]
@theazgra  They can be browsed, but not downloaded when credentials are setup.

**Comment by tomasfil** [comment]
But once you insert credentials for download, then your browser remembers it

**Comment by theazgra** [comment]
@tomasfil Ah i See, ok, that makes sense. :) this + path deny in HAProxy is sufficient for our needs. Thanks

**Pull request merged**
